### ISSUE TITLE
google_project policy_data breaking changes in v0.8.5

### ISSUE THREAD

**Issue by craigfurman** [hashicorp/terraform#11556]
# Synopsis
With an existing terraform-managed environment on GCP, switching to v0.8.5 when we are using existing configuration that contains a `policy_data` field in a `google_project` resource, the environment's IAM configuration ends up getting overwritten, not merged.

This constitutes a breaking change which deviates from behavior documented in v0.8.4. We know the field is deprecated, but the semantic change was not made clear.

### Terraform Version
Terraform v0.8.5

### Affected Resource(s)
- google_project

### Terraform Configuration Files
```hcl
# This was first `terraform apply`'d using v0.8.4
resource "google_project" "project" {
    id = "some-project"
    policy_data = "${data.google_iam_policy.opsman.policy_data}"
}

resource "google_service_account" "opsman" {
    account_id = "some-account-id"
    display_name = "opsman"
}

data "google_iam_policy" "opsman" {
  binding {
    role = "roles/iam.serviceAccountActor"
    members = [
      "serviceAccount:${google_service_account.opsman.email}",
    ]
  }

  binding {
    role = "roles/compute.instanceAdmin"
    members = [
      "serviceAccount:${google_service_account.opsman.email}",
    ]
  }

  binding {
    role = "roles/compute.networkAdmin"
    members = [
      "serviceAccount:${google_service_account.opsman.email}",
    ]
  }

  binding {
    role = "roles/compute.storageAdmin"
    members = [
      "serviceAccount:${google_service_account.opsman.email}",
    ]
  }
}
```

### Debug Output
We haven't reproduced this to get debug output without locking ourselves out of our GCP project (again), as overwriting the IAM policy deleted all of our project owners, and Google support had to bail us out.

### Panic Output
It didn't panic.

### Expected Behavior
`policy_data` is merged, with the semantic equivalent of `authoritative=false` in the new `google_project_iam_policy` resource.

### Actual Behavior
`policy_data` is overwritten (mirroring `authoritative=true`).

This is really a documentation/versioning bug. The line "The policy will be merged with any existing policy applied to the project." was removed from the `policy_data` field in `google_project`, but we feel that because the semantics have changed so much this should be highlighted.

### Steps to Reproduce
*Using the above hcl config*
1. `terraform apply` using v0.8.4
2. `terraform apply` using v0.8.5
3. Your GCP IAM config will be overwritten

### Important Factoids
Be careful to leave yourself one project owner when reproducing, or you'll lock yourself out of GCP. You can do this by having one defined in the TF config above.

cc @landesherr

**Comment by danawillow** [comment]
@evandbrown fyi in case you haven't seen

**Comment by evandbrown** [comment]
_This is an incomplete assessment of the problem, but I'm recording what I'm finding as I go_:

This was clearly introduced in 0.8.5, apparently by the addition of recording the project's IAM policy data in the `Read` lifecycle event. This triggers a diff where intermediate state has the actual project policy and config has the policy subset. `Update` compares state and config, believes bindings in state were intentionally removed (based on their absence in the config), and then removes them when it merges the two states. 

What I'm working on now:

1. How do we reconcile the new "import existing policy" behavior with configs that were migrated from v0.8.4?
2. Why did integration tests not catch this?

Commencing work now...

**Comment by evandbrown** [comment]
@paddyforan when working with a `*schema.ResourceData` is it possible to request the value of a key specifically from `*terraform.ResourceConfig`? I believe the default behavior of `Get()` merges config and state. Detecting if a change was present only in config and not state would be immensely useful here.

**Comment by allandrick** [comment]
@evandbrown,

Some additional data points that may be useful here - I too was doing this the old way like @craigfurman and when I updated to 0.8.5 and updated the syntax for google_project_iam_policy, I too lost access to the project I was working on, despite specifying "authortative = false" in my configuration.  Thankfully I am using the org.node and was able to re-add my Terraform service account to the specific project and start clean using the 0.8.5 syntax - however, things are still not quite right.

I use a Terraform module in all projects to add the default policy bindings I want - this works.  I then use local policy bindings and another google_project_iam_policy resource, also with authorative = false, in the same project to add custom roles to the project that aren't needed globally in the module.

This causes terraform to flip-flop between re-writing the policy bindings, especially where I am trying to add a specific IAM role to existing groups or service accounts that were defined with the base policy bindings for the project.

I am happy to share screengrabs if you need them, please DM.

Regards,

Leigh

**Comment by allandrick** [comment]
The fundamental issue seems to be that, regardless of which Terraform resource object is acting on the google_project_iam_policy, Terraform is removing existing roles from IAM users that don't match what is in the local Terraform file or state - and it explicitly shouldn't be doing this according to the documentation.

**Comment by allandrick** [comment]
Example:

in module "iam-account", we create:

```
resource "google_project_iam_policy" "google-project" {
    project = "${var.gcp_project}"
    policy_data = "${data.google_iam_policy.policy-bindings.policy_data}"
    authoritative = false
}

data "google_iam_policy" "policy-bindings" {
    # App Engine
    binding {
        role = "roles/appengine.appAdmin"
        members = [
            "group:${var.admins_group}",
        ]
    }
    binding {
        role = "roles/appengine.deployer"
        members = [
            "group:${var.users_group}",
        ]
    }
```

and in a local terraform file, we later create:

```
resource "google_service_account" "gcr-readonly" {
    account_id = "gcr-readonly"
    display_name = "Google Container Registry read-only service account"
}

data "google_iam_policy" "local-policy-bindings" {
  binding {
    role = "roles/container.viewer"
    members = [
      "serviceAccount:${google_service_account.gcr-readonly.email}",
    ]
  }
}

resource "google_project_iam_policy" "google-project" {
    project = "${var.gcp-project}"
    policy_data = "${data.google_iam_policy.local-policy-bindings.policy_data}"
    authoritative = false
}
```

this should only add this new policy binding to the Project IAM - and terraform plan says that's all it wants to do:

```
~ google_project_iam_policy.google-project
    policy_data: "{}" => "{\"bindings\":[{\"members\":[\"serviceAccount:gcr-readonly@xxxx-xxxx-xxxx.iam.gserviceaccount.com\"],\"role\":\"roles/container.viewer\"}]}"
```

So we apply it:

```
google_project_iam_policy.google-project: Modifying...
  policy_data: "{}" => "{\"bindings\":[{\"members\":[\"serviceAccount:gcr-readonly@xxxx-xxxx-xxxx.iam.gserviceaccount.com\"],\"role\":\"roles/container.viewer\"}]}"
google_project_iam_policy.google-project: Modifications complete
```

But then immediately a new terraform plan says it wants to change things again, because reality doesn't match what is in the module's state, even though it shouldn't be removing things that are already in IAM...

```
~ module.iam-account.google_project_iam_policy.google-project
    policy_data: "{\"bindings\":[{\"members\":[\"group:admins@group.company.com\"],\"role\":\"roles/appengine.appAdmin\"},{\"members\":[\"group:users@group.company.com\"],\"role\":\"roles/appengine.deployer\"},{\"members\":[\"serviceAccount:gcr-readonly@xxxx-xxxx-xxxx.gserviceaccount.com\"],\"role\":\"roles/container.viewer\"}]}" => "{\"bindings\":[{\"members\":[\"group:admins@group.company.com\"],\"role\":\"roles/appengine.appAdmin\"},{\"members\":[\"group:users@group.company.com\"],\"role\":\"roles/appengine.deployer\"}]}"
```

**Comment by evandbrown** [comment]
@allandrick Thanks for chiming in here! This current issue is specific to behavior in `google_project`, but it sounds like what you're seeing is with `google_project_iam_policy`. Would you be so kind as to open a new issue and paste in what you reported above, please? I'll start digging into what you've reported in the meantime, and thank you for the descriptive report.

**Comment by paddyforan** [comment]
So, first of all, apologies @craigfurman. You're entirely right that this is a backwards-incompatible change, and it's decidedly an unintentional one; I'm sorry for letting this slip through. We ran a lot of tests to avoid bugs like this, but a combination of a long-lived and large PR and numerous changes over time caused this scenario to go untested.

@evandbrown and I just were on a Hangout getting to the bottom of this, and here's my understanding of the situation:

* in 0.8.4, the `policy_data` was never refreshed in state; it was set by the data source, and never updated.
* In 0.8.5, the tests for `policy_data` were moved to being tests for `google_project_iam_policy` instead of duplicated, leaving a hole in our test coverage for ensuring that policies are merged correctly.
* While working on 0.8.5, we noticed `policy_data` wasn't being refreshed, and fixed it in `google_project_iam_policy`, then fixed it in `google_project` as well, to be consistent.
* The old test would have caught the merge problem, but as it had been moved to a different resource, they were missed.
* Now that we're setting the `policy_data` on refresh from the server, it gets stored in the state, and terraform thinks it owns those policies, even though they've never appeared in a config.
* In the `google_project_iam_policy` case, this isn't a problem, because when `authoritative` is false, we store the project IAM that doesn't exist in config at project creation in a `restore_data` property, and when setting the IAM, we merge the `restore_data` and the config.
* In `google_project`, we don't store `restore_data`, so we are in essence always considering it authoritative, and anything not in the config gets blown away.

I thought about it, and we essentially have three options:

1. We can essentially restore the old behaviour, where `policy_data` is never refreshed in state, and is always just the output of the data source. This will leave the attribute broken, in that it won't detect drift, but will restore the behaviour users have come to expect from it.
2. We can add an authoritative flag, much like `google_project_iam_policy` has, and essentially duplicate its behaviour. This will share the same semantics then, ideally, and will fix the broken old behaviour, but may come with some new edge cases and may still surprise users, as it's technically behaving in a new way, even if the semantics are the same.
3. We can snapshot the IAM policy on the creation of the resource and save the diff between it and the config in `restore_data`, the way we do for `google_project_iam_policy`. This is essentially the same as option 2, we just don't add a flag to override the behaviour. There is room for some unexpectedness here, however; I'm not sure how this interacts with import, for example.

In the interest of getting a fix out that works consistently with previous behaviour, I'll go ahead and do a quick PR that takes approach 1; it leaves drift-detection broken, but that seems like the lesser of two evils for a deprecated attribute, anyways. I'll also restore the test that would have caught this drift.

Again, apologies for the breakage; this is the exact scenario we want to avoid.

**Comment by evandbrown** [comment]
@paddyforan great summary; thanks! Feel free to piggy-back on #11728

**Comment by paddyforan** [comment]
Ah! Sorry @evandbrown, I was flying, and didn't get through my email backlog before commenting here, so I missed that. I'll definitely take a look at that. Thanks for taking point on it :)

**Comment by craigfurman** [comment]
Thanks for the detailed explanation @paddyforan!

Option does seem like a valid fix, especially as the attribute is deprecated and any oddness with not detecting drift won't be a problem for much longer (presumably).

For what it's worth, our team always found `google_project` a bit of an unusual resource to work with (in the context of IAM policy manipulation), in that it doesn't represent a new thing created by and then managed by terraform, but instead represents modification of a resource that already exists. I'm not sure there's a more "Typically Terraform" way of implementing this though.

**Comment by paddyforan** [comment]
Hey @craigfurman. This should be fixed now, and will be part of 0.8.6, which just landed.

Again, apologies for the backwards incompatibility--we really strive to avoid those, and are pretty serious about not including them except in major (0.X) releases. Thank you for the excellent issue report, though! Also, I would _love_ to hear more about your thoughts on how `google_project` is unusual to work with; I'm focusing some effort on that, and input is always appreciated. Feel free to open an issue (if you think the differences are something that should be fixed) or drop an email to paddy@hashicorp if you aren't sure whether it should result in changes or not.

Thanks!

**Comment by craigfurman** [comment]
No need for apologies @paddyforan , my colleagues and I are all appreciative of terraform, so thanks! We'll try out 0.8.6.

I'll find some time to organise my thoughts about google_project/google_project_iam_policy and drop you an email, as they're really product concerns and therefore I'm not sure if they should result in changes or not. It'll be good to open the discussion though.
